Dev

hsxyl opened the pull request.

hsxyl committed:
finish main functions.

hsxyl committed:
feat: change pool rule which impact how to control pool and convert

- A pool will has its onwer.Only onwer can deposit token into pool or withdraw from pool.
- Change pool rate storage form,and it also impacts calculating procss
  of converting.

hsxyl committed:
refactor: remove three useless view interface

- remove `fn minimum_received` in `trait ConvertorViewer`.
- remove `fn select_best_pool` in `trait ConvertorViewer`.
- remove `fn get_pools_by_token_direction` in `trait ConvertorViewer`.

hsxyl committed:
style: format code style by:cargo fmt and change edition into 2018

- The near simulator has conflicts with 2022 edition,so change it into
  2018.
- Use `cargo fmt` in `build.sh` to format codes.

hsxyl committed:
fix: fix withdraw_out_token mistake and build.sh path mistake

hsxyl committed:
build: adjust build.sh and scripts path

hsxyl committed:
feat: pool can set a deposit amount when be created and delete

hsxyl committed:
feat: add view method to get all pools of a creator

hsxyl committed:
test: simulation test of admin actions,creator actions,convertor actions

hsxyl committed:
feat: add account withdraw function

hsxyl committed:
feat: add two view methods: get_account_storage_debt and get_account

hsxyl committed:
feat: add functions for pausing contract by admin

riversyang commented on nep141-token-convertor-contract/src/account.rs:
```diff
@@ -0,0 +1,181 @@
+    }
+
+    pub fn storage_usage(&self) -> u64 {
+        INIT_ACCOUNT_STORAGE + self.tokens.len() as u64 * (ACC_ID_AS_KEY_STORAGE + U128_STORAGE)
```
I think the storage usage should be `U128_STORAGE +  + self.tokens.len() as u64 * (ACC_ID_AS_KEY_STORAGE + U128_STORAGE)`.

hsxyl replied:
I think both `U128_STORAGE` and  `INIT_ACCOUNT_STORAGE `  are ok.Just make sure size of increasing storage when adding an account  equals size of decreasing when remove an account.

riversyang commented on nep141-token-convertor-contract/src/token_receiver.rs:
```diff
@@ -0,0 +1,130 @@
+                );
+                let (receive_token_id, receive_token_amount) =
+                    self.internal_convert(convert_action.pool_id, &token_id, amount.0);
+                self.internal_send_tokens(&sender_id, &receive_token_id, receive_token_amount);
```
The function `internal_send_tokens` needs 30T gas to complete. But the prepaid gas for function `ft_transfer_call` is set to 25T in fungible token contract template. We need to double check whether the function `ft_transfer_resolved` can be executed successfully.

hsxyl replied:
I have add a check of gas in `internal_send_token` which requiring remind gas must more than `ft_transfer` and `ft_transfer_resolved` cost.

riversyang commented on nep141-token-convertor-contract/src/conversion_pool.rs:
```diff
@@ -0,0 +1,313 @@
+    }
+
+    #[payable]
+    fn delete_pool(&mut self, pool_id: PoolId) {
```
There are 3 possible promises which can be created in this function. I'm not sure what will happen if the prepaid gas can only fulfill the first one or two promise(s). Need to double check.
I recommend to check the prepaid gas before actual process.

hsxyl replied:
I have changed logic in `delete_pool`.Now requiring withdraw all `in_token` and `out_token` in pool before deleting pool.And admin can also execute withdraw token for creator now.

riversyang reviewed (changes requested):
I pointed out a few tricky parts in contract. And the tests failed in my computer because of compile errors. Need to double check.

hsxyl committed:
docs: correct some annotations

hsxyl committed:
refactor: refactor storage management

- Remove storage debt definition.
- Change storage_balance_bounds,it will calculate min bound dynamically.
- When saving account, it will check if near for storage is enough.
- When send nep141 token,it will check if user satisfy
  StorageBalanceBounds.min.

hsxyl committed:
refactor: rename get_storage_fee_need_deposit to get_storage_fee_gap_of

riversyang committed:
Merge pull request #3 from octopus-network/feat-prepay-storage-fee

refactor: refactor storage management

hsxyl committed:
fix: test and fix some storage bugs

- Storage_balance_bounds is a view method and it didn't allow use env::predecessor_account_id() when view it.So use internal_get_storage_balance_min_bound replace this method.
- Changing storage_deposit logic.
- When ft_on_transfer executing,the env::predecessor_account_id is ft_token id,so adding a param `account_id` into assert_storage_balance_bound_min.

riversyang committed:
Fix a bug and rename a few functions.

riversyang committed:
Merge pull request #4 from octopus-network/fix-cargo-test-fail

fix: test and fix some storage bugs

hsxyl committed:
refactor: refactor delete pool and withdraw token from pool

- Changing param: amount type from U128 into Option<U128> in `withdraw_token_in_pool`,if amount is Option::None,it means withdraw all tokens.
- Changing check logic in `withdraw_token_in_pool`,now it allows admin withdraw token.
- Changing `delete_pool` logic,requiring withdraw all in_token and out_token in pool before deleting pool.
- Adding a check of gas when `internal_send_token`.Requring remind gas must more than gas of `ft_transfer` and `ft_transfer_resolved` need.

riversyang committed:
Merge pull request #5 from octopus-network/feat-refact-delete-withdraw

refactor: refactor delete pool and withdraw token from pool
